go/libkb/key_pseudonym.go

```diff
@@ -0,0 +1,286 @@
+		// Compute the pseudonym
+		var pnym KeyPseudonym
+		pnym, err = MakeKeyPseudonym(info)
+		if err != nil {
+			return
+		}
+		(*pnymInfos)[i].KeyPseudonym = pnym
+		pnymReqs = append(pnymReqs, keyPseudonymReq{
+			Pseudonym:   pnym.String(),
+			ID:          info.ID.String(),
+			Application: int(info.Application),
+			KeyGen:      int(info.KeyGen),
+			Nonce:       hex.EncodeToString(info.Nonce[:]),
+		})
+	}
+
+	payload := make(JSONPayload)
+	payload["key_pseudonyms"] = pnymReqs
+
+	_, err = m.G().API.PostJSON(APIArg{
+		MetaContext: m,
+		Endpoint:    "team/key_pseudonym",
+		JSONPayload: payload,
+		SessionType: APISessionTypeREQUIRED,
+	})
+	if err != nil {
+		return
+	}
+	return nil
+}
+
+// GetKeyPseudonyms fetches info for a list of pseudonyms.
+// The output structs are returned in the order corresponding to the inputs.
+// The top-level error is filled if the entire request fails.
+// The error in each of the returned structs may be filled for per-pseudonym errors.
+func GetKeyPseudonyms(m MetaContext, pnyms []KeyPseudonym) ([]KeyPseudonymOrError, error) {
+	var pnymStrings []string
+	for _, x := range pnyms {
+		pnymStrings = append(pnymStrings, x.String())
+	}
+
+	var res getKeyPseudonymsRes
+	err := m.G().API.GetDecode(
+		APIArg{
+			MetaContext: m,
+			Endpoint:    "team/key_pseudonym",
+			SessionType: APISessionTypeREQUIRED,
+			Args: HTTPArgs{
+				"key_pseudonyms": S{Val: strings.Join(pnymStrings, ",")},
+			},
+		},
+		&res)
+	if err != nil {
+		return nil, err
+	}
+
+	// Validate the response
+	if len(res.KeyPseudonyms) != len(pnyms) {
+		return nil, &KeyPseudonymGetError{fmt.Sprintf("invalid server response for pseudonym get: len %v != %v",
+			len(res.KeyPseudonyms), len(pnyms))}
+	}
+	var resList []KeyPseudonymOrError
+	for i, received := range res.KeyPseudonyms {
+		resList = append(resList, checkAndConvertKeyPseudonymFromServer(pnyms[i], received))
+	}
+
+	return resList, nil
+}
+
+func checkAndConvertKeyPseudonymFromServer(req KeyPseudonym, received getKeyPseudonymRes) KeyPseudonymOrError {
+	mkErr := func(err error) KeyPseudonymOrError {
+		return KeyPseudonymOrError{
+			Info: nil,
+			Err:  err,
+		}
+	}
+
+	x := KeyPseudonymOrError{}
+
+	// This check is necessary because of sneaky typed nil.
+	// received.Err's type is lower than x.Err
+	// So doing `x.Err = received.Err` is bad if received.Err is nil.
+	// https://golang.org/doc/faq#nil_error
+	// https://play.golang.org/p/BnjVTGh-gO
+	if received.Err != nil {
+		x.Err = received.Err
+	}
+
+	if received.Info != nil {
+		info := KeyPseudonymInfo{}
+
+		info.ID = keybase1.UserOrTeamID(received.Info.ID)
+		if len(info.ID) != keybase1.TEAMID_LEN*2 { // hex encoded, 2 chars per byte
+			return mkErr(fmt.Errorf("user_or_team_id wrong length"))
+		}
+
+		info.KeyGen = KeyGen(received.Info.KeyGen)
+		info.Application = keybase1.TeamApplication(received.Info.Application)
+
+		n, err := hex.Decode(info.Nonce[:], []byte(received.Info.Nonce))
```

Comment:
see the comments above for doing this `keybase1/extras` as part of the decoding process (rather than by hand)